plugin/percona-pam-for-mysql/src/groups.c

```diff
@@ -41,23 +43,58 @@ struct groups_iter *groups_iter_new(const char *user_name)
   int error;
   struct groups_iter *it;
 
-  it= calloc(1, sizeof(struct groups_iter));
-  if (it == NULL)
-    return NULL;
+  if (gr_buf_size <= 0)
+  {
+    long gr_size_max, pw_size_max;
+    gr_size_max= sysconf(_SC_GETGR_R_SIZE_MAX);
+    pw_size_max= sysconf(_SC_GETPW_R_SIZE_MAX);
+    gr_buf_size= gr_size_max > pw_size_max ? gr_size_max : pw_size_max;
+  }
+
+  it= (struct groups_iter *) my_malloc(sizeof(struct groups_iter),
+                                       MYF(MY_FAE | MY_ZEROFILL));
+
+  it->buf_size= gr_buf_size;
+  if (it->buf_size <= 0)
+    it->buf_size= 1024;
+
+  it->buf= (char *) my_malloc(it->buf_size, MYF(MY_FAE));
 
-  error= getpwnam_r(user_name, &pwd, it->buf, max_nss_name_len, &pwd_result);
+  while ((error= getpwnam_r(user_name, &pwd, it->buf, it->buf_size,
+                            &pwd_result)) == ERANGE)
+  {
+    it->buf_size= it->buf_size * 2;
+    it->buf= (char *) my_realloc(it->buf, it->buf_size, MYF(MY_FAE));
+  }
   if (error != 0 || pwd_result == NULL)
   {
-    free(it);
+    fprintf(stderr, "auth_pam: Unable to obtain the passwd entry for the user "
+                    "'%s'.", user_name);
+    my_free(it->buf);
+    my_free(it);
     return NULL;
   }
 
-  it->ngroups= max_number_of_groups;
+  gr_buf_size= it->buf_size;
+
+  it->ngroups= 1024;
+  it->groups= (gid_t *) my_malloc(it->ngroups * sizeof(gid_t), MYF(MY_FAE));
   error= getgrouplist(user_name, pwd_result->pw_gid, it->groups, &it->ngroups);
   if (error == -1)
   {
-    free(it);
-    return NULL;
+    it->groups= (gid_t *) my_realloc(it->groups, it->ngroups * sizeof(gid_t),
+                                     MYF(MY_FAE));
+    error= getgrouplist(user_name, pwd_result->pw_gid, it->groups,
+                        &it->ngroups);
+    if (error == -1)
+    {
+      fprintf(stderr, "auth_pam: Unable to obtain the group access list for "
+                      "the user '%s'.", user_name);
+      my_free(it->buf);
```

Comment:
Loop doesn't work on mac unfortunately. As I said earlier, `getgrouplist` doesn't return -1 when I pass 4, but the user is the member of 15 groups... Here is an example:
```
$ ./aaa sergei 2
getgrouplist() returned -1; ngroups = 2

$ ./aaa sergei 3
ngroups = 3
20 (staff)
401 (com.apple.sharepoint.group.1)
12 (everyone)

$ ./aaa sergei 4
ngroups = 4
20 (staff)
401 (com.apple.sharepoint.group.1)
12 (everyone)
61 (localaccounts)

$ ./aaa sergei 5
ngroups = 5
20 (staff)
401 (com.apple.sharepoint.group.1)
12 (everyone)
61 (localaccounts)
79 (_appserverusr)

$ ./aaa sergei 10
ngroups = 10
20 (staff)
401 (com.apple.sharepoint.group.1)
12 (everyone)
61 (localaccounts)
79 (_appserverusr)
80 (admin)
81 (_appserveradm)
98 (_lpadmin)
33 (_appstore)
100 (_lpoperator)

$ ./aaa sergei 20
ngroups = 14
20 (staff)
401 (com.apple.sharepoint.group.1)
12 (everyone)
61 (localaccounts)
79 (_appserverusr)
80 (admin)
81 (_appserveradm)
98 (_lpadmin)
33 (_appstore)
100 (_lpoperator)
204 (_developer)
395 (com.apple.access_ftp)
398 (com.apple.access_screensharing)
399 (com.apple.access_ssh)
```
Moreover, `sysconf(_SC_NGROUPS_MAX)` returns 16 which in fact can be exceeded... Looks like we have no options on OS X if we are using `getgrouplist`.